File: android/airship-framework-proxy/src/main/java/com/urbanairship/android/framework/proxy/EmailRegistrationProxyOptions.kt
package com.urbanairship.android.framework.proxy

import com.urbanairship.contacts.EmailRegistrationOptions
import com.urbanairship.json.JsonMap
import java.util.Date

public data class EmailRegistrationProxyOptions(
    public val transactionalOptedIn: Long?,
    public val commercialOptedIn: Long?,
    public val properties: JsonMap?,
    public val doubleOptIn: Boolean
) {
    public constructor(json: JsonMap) : this(
        transactionalOptedIn = if (json.opt("transactionalOptedIn").isNull) null else json.opt("transactionalOptedIn").getLong(0),
        commercialOptedIn = if (json.opt("commercialOptedIn").isNull) null else json.opt("commercialOptedIn").getLong(0),
        properties = json.opt("properties").map,
        doubleOptIn = json.opt("doubleOptIn").getBoolean(false)
    )

    public fun toEmailRegistrationOptions(): EmailRegistrationOptions {
        val transactionalDate = transactionalOptedIn?.let { Date(it) }
        val commercialDate = commercialOptedIn?.let { Date(it) }

        return if (commercialDate != null) {
            EmailRegistrationOptions.commercialOptions(
                commercialDate,
                transactionalDate,
                properties
            )
        } else {
            EmailRegistrationOptions.options(
                transactionalDate,
                properties,
                doubleOptIn
            )
        }
    }
}

File: ios/AirshipFrameworkProxy/EmailRegistrationProxyOptions.swift
/* Copyright Airship and Contributors */

import Foundation

#if canImport(AirshipKit)
import AirshipKit
#elseif canImport(AirshipCore)
import AirshipCore
#endif

public struct EmailRegistrationProxyOptions: Decodable, Sendable {
    let transactionalOptedInMilliseconds: Double?
    let commercialOptedInMilliseconds: Double?
    let properties: [String: String]?
    let doubleOptIn: Bool

    private enum CodingKeys: String, CodingKey {
        case transactionalOptedInMilliseconds = "transactionalOptedIn"
        case commercialOptedInMilliseconds = "commercialOptedIn"
        case properties
        case doubleOptIn
    }

    func toEmailRegistrationOptions() -> EmailRegistrationOptions {
        let transactionalDate = transactionalOptedInMilliseconds.map {
            Date(timeIntervalSince1970: $0 / 1000.0)
        }
        let commercialDate = commercialOptedInMilliseconds.map {
            Date(timeIntervalSince1970: $0 / 1000.0)
        }

        if commercialDate != nil {
            return EmailRegistrationOptions.commercialOptions(
                transactionalOptedIn: transactionalDate,
                commercialOptedIn: commercialDate,
                properties: properties
            )
        } else {
            return EmailRegistrationOptions.options(
                transactionalOptedIn: transactionalDate,
                properties: properties,
                doubleOptIn: doubleOptIn
            )
        }
    }
}
